pkg/cloud/affinity_groups_test.go

```diff
@@ -48,22 +55,79 @@ var _ = Describe("AffinityGroup Unit Tests", func() {
 		mockCtrl.Finish()
 	})
 
-	It("fetches an affinity group", func() {
-		dummies.AffinityGroup.ID = "" // Force name fetching.
-		ags.EXPECT().GetAffinityGroupByName(dummies.AffinityGroup.Name).Return(&cloudstack.AffinityGroup{}, 1, nil)
+	Context("Fetch or Create Affinity group", func() {
+		It("fetches an affinity group by Name", func() {
+			dummies.AffinityGroup.ID = "" // Force name fetching.
+			ags.EXPECT().GetAffinityGroupByName(dummies.AffinityGroup.Name).Return(&cloudstack.AffinityGroup{}, 1, nil)
+
+			Ω(client.GetOrCreateAffinityGroup(dummies.AffinityGroup)).Should(Succeed())
+		})
+
+		It("fetches an affinity group by ID", func() {
+			ags.EXPECT().GetAffinityGroupByID(dummies.AffinityGroup.ID).Return(&cloudstack.AffinityGroup{}, 1, nil)
+
+			Ω(client.GetOrCreateAffinityGroup(dummies.AffinityGroup)).Should(Succeed())
+		})
+
+		It("creates an affinity group", func() {
+			dummies.SetDummyDomainAndAccount()
+			dummies.SetDummyDomainID()
+			ags.EXPECT().GetAffinityGroupByID(dummies.AffinityGroup.ID).Return(nil, -1, fakeError)
+			ags.EXPECT().NewCreateAffinityGroupParams(dummies.AffinityGroup.Name, dummies.AffinityGroup.Type).
+				Return(&cloudstack.CreateAffinityGroupParams{})
+			ags.EXPECT().CreateAffinityGroup(ParamMatch(And(NameEquals(dummies.AffinityGroup.Name)))).
+				Return(&cloudstack.CreateAffinityGroupResponse{}, nil)
+
+			Ω(client.GetOrCreateAffinityGroup(dummies.AffinityGroup)).Should(Succeed())
+		})
+
+		It("creates an affinity group if Name provided returns more than one affinity group", func() {
```

Comment:
Now that I'm reading this, this behavior may not actually be what we want. This could be a good catch.

I think we probably want to return error if more than one affinity group comes back when listed by name?

We assumed the name to be unique.

Reply:
At least within account/domain/endpoint the affinity group name is enforced as unique.
![image](https://user-images.githubusercontent.com/93345136/184942063-69c5e518-2a61-4a2f-bcc8-46b2d971328f.png)

Reply:
So I think we should never really see this happen, but I think we'd actually prefer an error here.

Either way, this test is still fine to merge. It specs the current behavior.